# Домашне завдання №3 (2-1): Тестування стат.гіпотез

## Ольга Шапошник

In [2]:
import numpy as np
from statsmodels.stats.weightstats import ztest, ttest_ind

В цьому ДЗ потренуємось виконувати Z-тести для тестування гіпотез в різних задачах. Для коректної імплементації та інтерпретації тестів рекоемндую переглянути спершу документацію:

https://www.statsmodels.org/dev/generated/statsmodels.stats.weightstats.ztest.html

### Завдання 1: Z-тест для однієї вибірки

**Мета**: Перевірити, чи середнє значення вибірки відрізняється від гіпотетичного середнього.

**Припущення**: Середній ріст студентів університету вважається 170 см. Ми хочемо перевірити, чи середній ріст випадкової вибірки студентів істотно більший. Вибірка задана в змінній `heights`.

Для виконання завдання імпортуйте метод для проведення `ztest`, виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи справдилось припущення про середній зріст студентів на рівні значущості 0.05?

Використовуємо тут Z-тест, оскільки вибірка достатньо велика (більше 30 екземплярів) і можемо вважати, що її стандартне відхилення близьке до ст. відхилення генеральної сукупності, а значить - відоме.

In [3]:
heights = [174, 171, 175, 179, 170, 170, 179, 175, 169, 174, 169, 169, 173, 162, 163, 169, 166,
           173, 167, 164, 179, 170, 172, 164, 169, 175, 169, 174, 169, 169, 173, 162, 177]

In [4]:
mean_height = np.mean(heights)
mean_null = 170
alpha = 0.05
std = np.std(heights)
n = len(heights)
st = std / np.sqrt(n)
print(f"mean: {mean_height:.2f}, std: {std:.2f}, number of records: {n:.2f}")

mean: 170.70, std: 4.65, number of records: 33.00


In [5]:
z_test, p_value = ztest(heights, value=mean_null, alternative='larger')
print(f'Z-test: {z_test:.4f}, p-value: {p_value:.4f}')

Z-test: 0.8482, p-value: 0.1982


In [6]:
if p_value < alpha:
    print('H0 is rejected. Accept H1')
    print('Mean height is significantly greater than 170')
else:   
    print('H0 is not rejected. Accept H0')  
    print('Mean height is NOT significantly greater than 170')  

H0 is not rejected. Accept H0
Mean height is NOT significantly greater than 170


### Висновок ДЗ1
- Основна мета була Перевірити, чи середнє значення вибірки відрізняється від гіпотетичного середнього.
- Гіпотетичного середнє = 170
- Середнє для нашої вибірки 170.70 (при std = 4.65),
- Для того аби перевірити чи зріст у вибірка наших студентів статистично вищий ми маємо використати налаштування: alternative='larger', в такому випадку:
    - *Нульова гіпотеза*: Середнє вибірки дорівнює або менше за 170 см
    - *Альтернативна гіпотеза*: Середнє вибірки більше за 170 см
- На основі ztest ми маємо наступні розрахунки Z-test: 0.8482, p-value: 0.1982, при заданому рівню значущості у 0.05
- Розраховане p-value *більше* за заданий рівень значущості, тому ми маємо прийняти нульову гіпотезу. Яка каже що Середнє вибірки дорівнює або менше за 170 см
- на даному етапі в нас немає достатніх доказів, що середній зріст вибірки значно перевищує 170


### Завдання 2: Z-тест для двох незалежних вибірок

**Завдання 2.1.**

**Мета**: Перевірити, чи існує статистична різниця між середніми оцінками двох груп студентів.

**Припущення**: Група A отримала новий навчальний курс, а група B продовжувала зі стандартним курсом. Ми перевіряємо, чи є новий курс ефективнішим.

Виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи справдилось припущення щодо студентів на рівні значущості 0.05?

In [7]:
group_a_scores = [78.55, 72.25, 79.88, 75.  , 76.54, 74.99, 87.26, 77.93, 72.71,
       82.11, 71.9 , 79.04, 68.2 , 71.36, 78.98, 81.69, 78.86, 77.42,
       76.49, 70.61, 74.4 , 75.7 , 83.29, 79.72, 69.18, 79.62, 76.07,
       74.62, 81.06, 83.15, 82.66, 73.8 , 76.45, 79.66, 82.88, 75.6 ,
       77.07, 72.47, 72.02, 82.06]

group_b_scores = [81.78, 74.64, 80.02, 76.81, 71.77, 76.81, 82.69, 74.82, 82.82,
       61.9 , 79.11, 75.44, 73.5 , 75.46, 65.06, 73.9 , 76.79, 82.39,
       72.41, 70.96, 72.49, 79.58, 76.64, 72.35, 77.57, 75.49, 79.84,
       71.49, 73.36, 73.04, 67.68, 76.48, 76.31, 75.03, 73.83, 67.92,
       72.9 , 73.29, 70.99, 74.19]

In [8]:
mean_A = np.mean(group_a_scores)
mean_B = np.mean(group_b_scores)
std_A = np.std(group_a_scores)
std_B = np.std(group_b_scores)
n_A = len(group_a_scores)
n_B = len(group_b_scores)

alpha_2 = 0.05

print(f"mean A: {mean_A:.2f}, std A: {std_A:.2f}, number of records A: {n_A:.2f}")
print(f"mean B: {mean_B:.2f}, std B: {std_B:.2f}, number of records B: {n_B:.2f}")

mean A: 77.08, std A: 4.31, number of records A: 40.00
mean B: 74.74, std B: 4.46, number of records B: 40.00


In [9]:
z_test_2, p_value_2 = ztest(group_a_scores, group_b_scores, value=0, alternative='two-sided')
print(f'Z-test: {z_test_2:.4f}, p-value: {p_value_2:.4f}')

Z-test: 2.3574, p-value: 0.0184


In [10]:
if p_value_2 < alpha_2:
    print('H0 is rejected. Accept H1')
    print('Mean scores of group A and group B are significantly different')
else:
    print('H0 is not rejected. Accept H0')
    print('Mean scores of group A and group B are NOT significantly different')    

H0 is rejected. Accept H1
Mean scores of group A and group B are significantly different


In [22]:
z_test_2, p_value_2 = ztest(group_a_scores, group_b_scores, value=0, alternative='two-sided')
print(f'Z-test: {z_test_2:.4f}, p-value: {p_value_2:.4f}')

Z-test: 2.3574, p-value: 0.0184


### Висновок ДЗ 2_1
- Перевірити, чи існує статистична різниця між середніми оцінками двох груп студентів.
- При тому що статистика показує
    - для групи А: mean A: 77.08, std A: 4.31
    - для групи В: mean B: 74.74, std B: 4.46
- Для того аби перевірити відношення в двох вибірках сформуємо гіпотези:
    - **Нульова гіпотеза**: Середні значення для обох груп однакові
    - **Альтернативна гіпотеза**: Середні значення для груп відрізняються

- На основі ztest ми маємо наступні розрахунки Z-test: 2.3574, p-value: 0.0184, при заданому рівню значущості у 0.05

- Розраховане p-value *менше* за заданий рівень значущості, тому ми маємо відхилити нульову гіпотезу (та прийняти альтернативну). 
- Тому ми можемо стверджувати що існує статистично значущі відмінності між середніми оцінками групи A і групи B. 
- Для того аби дізнатися група А мають кращі результати можна використати (alternative='larger'). В нашому випадку ми бачимо на основі статистики що середнє значення для групи А (77.08) вищий за середнє значення в групі В (74.74). Отже новий курс є більш ефективним

**Завдання 2.2.**


**Мета**: Побачити, як впливає розмір вибірок на результат тесту.

**Завдання**: Уявіть, що з групи А ми маємо тільки перші 5 записів, а з групи В - усі. Таке могло статись, якщо ми вже маємо результати тестів учнів за попередньою програмою, а тести за новою програмою здали наразі лише 5 учнів і ми вирішили не чекати довше.
Виберіть перші 5 записів для групи А і проведіть t-test (для цього використовуємо наступний [метод](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_ind.html)). Виведіть значення p-value. Зробіть висновок, який результат ви бачите щодо стат. значущості на рівні 0.05 для цього експерименту?

Тут проводимо t-test оскільки одна з вибірок - дуже мала.

In [12]:
group_a_scores_short = group_a_scores[:5]

mean_A_short = np.mean(group_a_scores_short)
std_A_short = np.std(group_a_scores_short)
n_A_short = len(group_a_scores_short)

print(f"mean A short: {mean_A_short:.2f}, std A short: {std_A_short:.2f}, number of records A short: {n_A_short:.2f}")
print(f"mean B: {mean_B:.2f}, std B: {std_B:.2f}, number of records B: {n_B:.2f}")

mean A short: 76.44, std A short: 2.68, number of records A short: 5.00
mean B: 74.74, std B: 4.46, number of records B: 40.00


In [15]:
t_test = ttest_ind(group_a_scores_short, group_b_scores, alternative='two-sided')
p_value_ttest = t_test[1]
print(f'T-test: {t_test[0]:.4f}, p-value: {t_test[1]:.4f}')

T-test: 0.8168, p-value: 0.4185


In [16]:
if p_value_ttest < alpha_2:
    print('H0 is rejected. Accept H1')
    print('Mean scores of group A and group B are significantly different')
else:
    print('H0 is not rejected. Accept H0')
    print('Mean scores of group A and group B are NOT significantly different')    

H0 is not rejected. Accept H0
Mean scores of group A and group B are NOT significantly different


### Висновок ДЗ 2_2
- Нам необхідно зрозуміти як розмір вибірок впливає на значення тесту. 
- Статистика:
    - для групи А: mean = 76.44, std = 2.68, number of records = 5
    - для групи В: mean = 74.74, std = 4.46, number of records = 40
- Для того аби перевірити відношення в дво вибірках сформуємо гіпотези:
    - **Нульова гіпотеза**: Середні значення для обох груп однакові
    - **Альтернативна гіпотеза**: Середні значення для груп відрізняються
- Через те що ми маємо малу вибірку, то використовуємо t test (T-test: 0.8168, p-value: 0.4185)

- Розраховане p-value *Більше* за заданий рівень значущості (0.05), тому ми не можемо відхилити нульову гіпотезу.
- Таким чином при зменшенні к-ті записів в групі А ми більше не можемо стверджувати що ми маємо статистично значущу різницю між групами. 
- Тому навіть якщо існує реальна різниця між двома вибірками (що було в попередньому випадку), то при недостатній к-ті записів ми можемо отримати статистично не значущий результат



### Завдання 3\*: Z-тест для двох **пов'язаних** вибірок

**Мета**: Перевірити, чи вплинув тренінг на продуктивність працівників, коли всі справробітники проходили тренінг.

**Припущення**: Наші працівники проходять однаковий тренінг, і ми хочемо з'ясувати, чи покращилися їхні показники після тренінгу на рівні значущості 0.05. Виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи поліпшились показники працівників на рівні значущості 0.05?

Зверніть увагу, ці вибірки повʼязані між собою, значить не буде коректно виконати z-test між ними, натомість, ми можемо порівняти, чи різниця в результатах (продуктивності співробітників) є стат значущою.

In [17]:
before_training = [57.82, 37.63, 36.8 , 55.22, 52.97, 52.5 , 53.46, 43.2 , 52.32,
       52.93, 42.86, 68.66, 54.74, 38.09, 56.57, 40.25, 57.87, 61.59,
       41.79, 59.63, 54.13, 58.22, 68.97, 47.55, 42.46, 41.1 , 41.84,
       49.23, 53.41, 52.77]

after_training = [62.47, 40.66, 42.7 , 57.69, 61.41, 56.76, 54.75, 44.06, 56.29,
       55.48, 47.28, 72.6 , 57.59, 39.39, 56.54, 42.36, 62.58, 65.01,
       42.3 , 62.98, 57.9 , 59.45, 72.28, 50.66, 43.18, 44.82, 45.96,
       54.4 , 58.52, 53.01]

before = np.array(before_training)
after = np.array(after_training)
training_diff = after - before

training_diff

array([ 4.65,  3.03,  5.9 ,  2.47,  8.44,  4.26,  1.29,  0.86,  3.97,
        2.55,  4.42,  3.94,  2.85,  1.3 , -0.03,  2.11,  4.71,  3.42,
        0.51,  3.35,  3.77,  1.23,  3.31,  3.11,  0.72,  3.72,  4.12,
        5.17,  5.11,  0.24])

In [18]:
mean_before = np.mean(before)
mean_after = np.mean(after)
std_before = np.std(before)
std_after = np.std(after)

mean_trainung_diff = np.mean(training_diff)
mean_trainung_diff_null = 0 # no difference
std_trainung_diff = np.std(training_diff)
n_trainung_diff = len(training_diff)
alpha_3 = 0.05
print(f"mean diff: {mean_trainung_diff:.2f}, std  diff: {std_trainung_diff:.2f}, number of records: {n_trainung_diff:.2f}")
print(f"mean before: {mean_before:.2f}, std before: {std_before:.2f}")
print(f"mean after: {mean_after:.2f}, std after: {std_after:.2f}")

mean diff: 3.15, std  diff: 1.86, number of records: 30.00
mean before: 50.89, std before: 8.58
mean after: 54.04, std after: 8.96


In [19]:
z_test_3, p_value_3 = ztest(training_diff, value = mean_trainung_diff_null, alternative='larger')
print(f'Z-test: {z_test_3:.4f}, p-value: {p_value_3:.4f}')

Z-test: 9.1389, p-value: 0.0000


In [20]:
if p_value_3 < alpha_3:
    print('H0 is rejected. Accept H1')
    print('Mean difference is significantly greater than 0')
else:   
    print('H0 is not rejected. Accept H0')  
    print('Mean difference is NOT significantly greater than 0')    

H0 is rejected. Accept H1
Mean difference is significantly greater than 0


### Висновок ДЗ 3
- Нам необхідно Перевірити, чи вплинув тренінг на продуктивність працівників (чи поліпшились показники працівників).
- Через те що ці дві події залежні одна від одної ми маємо перефразувати задачу, аби позбутися цього обмеження. Для цього розрахуємо різницю між продуктивність Після тренінгу та До тренінгу (аби побачити чи є різниця між поточним станом і минулим))
- Статистика:
    - До тренінгу: mean = 50.89, std = 8.58
    - Після тренінгу: mean = 54.04, std = 8.96
    - Різниця: mean = 3.15, std  = 1.86
- Гіпотетичне середнє = 0 (припущення що різниця між До та Після немає, показники залишилися однаковими)
- Для того аби перевірити чи показники статистично поліпшилися оберемо налаштування: alternative='larger'
    - **Нульова гіпотеза**: Середня різниця продуктивності між показниками дорівнює\менше 0 
    - **Альтернативна гіпотеза**: Середня різниця продуктивності між показниками більша за 0


- На основі ztest ми маємо наступні розрахунки Z-test: 9.1389, p-value: 0.0000, при заданому рівню значущості у 0.05
- Розраховане p-value *менше* за заданий рівень значущості, тому ми маємо відхилити нульову гіпотезу (та прийняти альтернативну). 
- Як висновок ми приймаємо що різниця між продуктивністю До та Після тренінгу поліпшилася. Тренінг статистично значимо вплинув на продуктивність а позитивну сторону.

